### ISSUE TITLE
ddtrace/tracer: network latency can cause blocking and lost traces

### ISSUE THREAD

**Issue by gbbr** [DataDog/dd-trace-go#465]
Due to the synchronous flushing nature of the tracer, when network latency is present, it is possible that the extra wait time will block the [the loop](https://github.com/outdoorsy/dd-trace-go/blob/4510a511ad295079fc65375b19864abc240a9d23/ddtrace/tracer/tracer.go#L149-L166) for too long, causing lost traces. This was reported multiple times by users.

```
2019/06/25 14:19:19 Datadog Tracer Error (v1.10.0): lost traces (count: 5), error: payload queue full, dropping trace (repeated 201 times)
```

We should make it so that the queue is flushed asynchronously.

**Comment by jpcope** [comment]
To clarify for future observers:

If you see the message `"payload queue full, dropping trace"`
or something to that effect being emitted by the `pushTrace` function at https://github.com/DataDog/dd-trace-go/blob/746d442e156a14c4d1d71c24757084ec28dad5d5/ddtrace/tracer/tracer.go#L186

It does not mean that your application is experiencing increased latency due to explicitly blocking on this call. The trace handoff actually occurs asynchronously, this log message is stating that the asynchronous handoff failed and the payload for that trace will be dropped. You may experience increased latency in your instrumented application due to extra logs, or decreased network throughput.

**Comment by uudashr** [comment]
Do we have fix or conclusion on this?

I'm experiencing the same here.

**Comment by gbbr** [comment]
@uudashr which tracer version are you using? The issue that this post describes has already been fixed. I'm going to close it but we can discus your specific issue either here or in a new issue if you wish to open one.

The asynchronous flushing of the queue was implemented in #549
